rasa/nlu/featurizers/featurizer.py

```diff
@@ -1,48 +1,142 @@
+Featurizer = Featurizer
+SparseFeaturizer = SparseFeaturizer
+DenseFeaturizer = DenseFeaturizer
 
-        # makes sure the alias name is set
-        # Necessary for `unique_name` to be defined
-        self.component_config = component_config
-        component_config.setdefault(FEATURIZER_CLASS_ALIAS, self.unique_name)
+FeatureType = TypeVar("FeatureType")
 
-        super().__init__(component_config)
 
+class Featurizer2(Generic[FeatureType]):
+    """Base class for all featurizers."""
 
-class DenseFeaturizer(Featurizer):
     @staticmethod
-    def _calculate_sentence_features(
-        features: np.ndarray, pooling_operation: Text
-    ) -> np.ndarray:
-        # take only non zeros feature vectors into account
-        non_zero_features = np.array([f for f in features if f.any()])
+    def get_default_config() -> Dict[Text, Any]:
+        """Returns the component's default config."""
+        return {FEATURIZER_CLASS_ALIAS: None}
 
-        # if features are all zero just return a vector with all zeros
-        if non_zero_features.size == 0:
-            return np.zeros([1, features.shape[-1]])
+    def __init__(self, config: Optional[Dict[Text, Any]] = None) -> None:
+        """Instantiates a new featurizer.
 
-        if pooling_operation == MEAN_POOLING:
-            return np.mean(non_zero_features, axis=0, keepdims=True)
+        Args:
+          config: configuration
+        """
+        super().__init__()
+        self.validate_config(config)
+        self._config = config
+        # TODO: we can't use COMPONENT_INDEX anymore to create a unique name
+        # on the fly --> take care of that in graph validation
+        self._identifier = self._config.get(FEATURIZER_CLASS_ALIAS, None)
 
-        if pooling_operation == MAX_POOLING:
-            return np.max(non_zero_features, axis=0, keepdims=True)
+    @property
+    def identifier(self) -> Text:
+        """Returns an identifier for this featurizer.
 
-        raise ValueError(
-            f"Invalid pooling operation specified. Available operations are "
-            f"'{MEAN_POOLING}' or '{MAX_POOLING}', but provided value is "
-            f"'{pooling_operation}'."
-        )
+        This name should be unique among all featurizers in the graph where this
+        component is used.
+        """
+        return self._identifier
 
+    # TODO: validate config for this component in the context of a recipe
 
-class SparseFeaturizer(Featurizer):
-    pass
+    @classmethod
+    def validate_config(cls, config: Dict[Text, Any]) -> None:
+        """Validates that the featurizer is configured properly."""
+        identifier = config.get(FEATURIZER_CLASS_ALIAS, [])
+        if not identifier:
+            raise ValueError(
+                f"Expected the config to map key `{FEATURIZER_CLASS_ALIAS}` "
+                f"to some name only found {config}"
+            )
+
+    def validate_compatibility_with_tokenizer(self, tokenizer: Tokenizer) -> None:
```

Comment:
Cool idea! Can we instead of the comment have an issue for this or put in the document with follow-up ideas?

Reply:
Added [here](https://www.notion.so/rasa/graph-Validation-static-check-of-configs-bfc65b6d97714fd099938cdbabca0f53) - Might also be included in some of the upcoming Recipe tasks, no? Cause some of the checks have been done in the version we're replacing here (most of all in the ensemble.py)

Reply:
Yep, can do 👍🏻